From d92a9857aa76b2206d78c5e70362c8e9eec2e1e9 Mon Sep 17 00:00:00 2001
From: Yann Dirson <yann.dirson@vates.tech>
Date: Mon, 3 Feb 2025 17:06:19 +0100
Subject: [PATCH] xenguest: activate nested virt when requested

---
 tools/xenguest/xenguest_stubs.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/tools/xenguest/xenguest_stubs.c b/tools/xenguest/xenguest_stubs.c
index 4eda3ee..20dc7ce 100644
--- a/tools/xenguest/xenguest_stubs.c
+++ b/tools/xenguest/xenguest_stubs.c
@@ -304,6 +304,11 @@ static int construct_cpuid_policy(const struct flags *f, bool hvm, bool restore)
     if ( !f->pae )
         clear_bit(X86_FEATURE_PAE, featureset);
 
+    if ( f->nested_virt ) {
+        set_bit(X86_FEATURE_VMX, featureset);
+        set_bit(X86_FEATURE_SVM, featureset);
+    }
+
     /*
      * Optionally advertise ITSC, given hardware support an a non-migratealbe
      * domain.
-- 
2.39.5

