diff --git a/xen/Makefile b/xen/Makefile
index b062e707f3..615785d936 100644
--- a/xen/Makefile
+++ b/xen/Makefile
@@ -393,6 +393,7 @@ CFLAGS-$(CONFIG_CC_SPLIT_SECTIONS) += -ffunction-sections -fdata-sections
 
 CFLAGS += -nostdinc -fno-builtin -fno-common
 CFLAGS += -Werror -Wredundant-decls -Wno-pointer-arith
+CFLAGS += -Werror=no-unused-variable
 CFLAGS += -Wdeclaration-after-statement
 $(call cc-option-add,CFLAGS,CC,-Wvla)
 CFLAGS += -pipe -D__XEN__ -include $(srctree)/include/xen/config.h
diff --git a/xen/common/pdx.c b/xen/common/pdx.c
index 480e873798..faa6dfd242 100644
--- a/xen/common/pdx.c
+++ b/xen/common/pdx.c
@@ -89,7 +89,7 @@ void set_pdx_range(unsigned long smfn, unsigned long emfn)
 #endif
 
 /* Generic PFN compression helpers. */
-static struct pfn_range {
+static __attribute__((unused)) struct pfn_range {
     unsigned long base_pfn, pages;
 } ranges[MAX_PFN_RANGES] __initdata;
 static unsigned int __initdata nr_ranges;
