diff --git a/xen/common/pdx.c b/xen/common/pdx.c
index 480e873798..21d72cfdeb 100644
--- a/xen/common/pdx.c
+++ b/xen/common/pdx.c
@@ -91,7 +91,7 @@ void set_pdx_range(unsigned long smfn, unsigned long emfn)
 /* Generic PFN compression helpers. */
 static struct pfn_range {
     unsigned long base_pfn, pages;
-} ranges[MAX_PFN_RANGES] __initdata;
+} ranges[MAX_PFN_RANGES] __initdata __attribute__((unused));
 static unsigned int __initdata nr_ranges;
 
 static bool __initdata pdx_compress = true;
