diff --git a/buildconfigs/config-debug b/buildconfigs/config-debug
index 36b8b77ae4..ac6ab86d4a 100644
--- a/buildconfigs/config-debug
+++ b/buildconfigs/config-debug
@@ -160,7 +160,7 @@ CONFIG_SYSCTL=y
 
 CONFIG_PM_OP=y
 CONFIG_PM_STATS=y
-CONFIG_LOCKDOWN_DEFAULT=y
+# CONFIG_LOCKDOWN_DEFAULT is not set
 # end of Common Features
 
 #
diff --git a/buildconfigs/config-release b/buildconfigs/config-release
index 644bd4e493..efdbf8bb11 100644
--- a/buildconfigs/config-release
+++ b/buildconfigs/config-release
@@ -160,7 +160,7 @@ CONFIG_SYSCTL=y
 
 CONFIG_PM_OP=y
 CONFIG_PM_STATS=y
-CONFIG_LOCKDOWN_DEFAULT=y
+# CONFIG_LOCKDOWN_DEFAULT is not set
 # end of Common Features
 
 #
